doc/source/whatsnew/v2.0.0.rst

```diff
@@ -1371,6 +1371,7 @@ Groupby/resample/rolling
 - Bug in :meth:`.DataFrameGroupBy.agg` with ``engine="numba"`` failing to respect ``as_index=False`` (:issue:`51228`)
 - Bug in :meth:`DataFrameGroupBy.agg`, :meth:`SeriesGroupBy.agg`, and :meth:`Resampler.agg` would ignore arguments when passed a list of functions (:issue:`50863`)
 - Bug in :meth:`DataFrameGroupBy.ohlc` ignoring ``as_index=False`` (:issue:`51413`)
+- Bug in :meth:`BaseGrouper.groups` when grouping by datetime, zip() was using :meth:`BaseGrouper.__iter__` that produced incorrect result in case of :class:`BinGrouper`. Fixed by adding :meth:`BinGrouper.__iter__` that returns correct iterable (:issue:`51158`)
```

Comment:
only need a description of the user-facing bug (i.e. `GroupBy.groups`), not the internals

Reply:
Thanks. Fixed all your comments.